lib/fog/azurerm/models/storage/file.rb

```diff
@@ -4,147 +4,386 @@ class AzureRM
+        #                          Content-Disposition Sets content disposition for the file. For exampple, 'attachment; filename=testing.txt'.
+        # @option options [Hash]   metadata            Sets custom metadata values to store with the file.
+        #
+        # @return [Boolean]
+        #
         def save(options = {})
-          requires :key
-          requires :directory
-          merge_attributes(options)
-          merge_attributes(File.parse(service.upload_block_blob_from_file(directory, key, file_path, options)))
+          update_body = options.delete(:update_body)
+          update_body = true if update_body.nil?
+          requires :directory, :key
+          raise ArgumentError.new('body is required when update_body is true') if update_body && attributes[:body].nil?
+
+          remap_attributes(
+            options,
+            'Blob-Type'           => :blob_type,
+            'Content-Type'        => :content_type,
+            'Content-Encoding'    => :content_encoding,
+            'Content-Language'    => :content_language,
+            'Cache-Control'       => :cache_control,
+            'Content-Disposition' => :content_disposition
+          )
+          options = {
+            blob_type: blob_type,
+            content_type: content_type,
+            content_encoding: content_encoding,
+            content_language: content_language,
+            cache_control: cache_control,
+            content_disposition: content_disposition,
+            metadata: metadata
+          }.merge!(options)
+          options = options.reject { |key, value| key == 'content_md5' || value.nil? || value.to_s.empty? }
+
+          if update_body
+            if options[:blob_type].nil? || options[:blob_type] == 'BlockBlob'
+              if Fog::Storage.get_body_size(body) <= SINGLE_BLOB_PUT_THRESHOLD
+                blob = service.create_block_blob(directory.key, key, body, options)
+              else
+                multipart_save_block_blob(options)
+                blob = service.get_blob_properties(directory.key, key)
+              end
+            else
+              blob_size = Fog::Storage.get_body_size(body)
+              raise "The page blob size must be aligned to a 512-byte boundary. But the file size is #{blob_size}." if (blob_size % 512).nonzero?
+
+              save_page_blob(blob_size, options)
+              blob = service.get_blob_properties(directory.key, key)
+            end
+
+            data = parse_storage_object(blob)
+            merge_attributes(data)
+            attributes[:content_length] = Fog::Storage.get_body_size(body)
+            attributes[:content_type] ||= Fog::Storage.get_content_type(body)
+          else
+            service.put_blob_metadata(directory.key, key, options[:metadata]) if options[:metadata]
+            options.delete(:metadata)
+            service.put_blob_properties(directory.key, key, options)
+
+            blob = service.get_blob_properties(directory.key, key)
```

Comment:
"save" method is too long , it can be divided into small methods, as line 97-110 is a logic part so it should be moved to a separate method and that method can be called here